joomla/joomla-cms
# Chunked downloads of update packages

Pull Request for Issue #36741 .

### Summary of Changes

You can now tell Joomla Update to download the upgrade packages one megabyte at a time. This allows the update to download on very slow servers (see the linked issue).

Tagging @zero-24 @laoneo @HLeithner 

### Why bother on Joomla 3?

First of all, the majority of Joomla sites are still on Joomla 3. It will take a long while for everyone to migrate to Joomla 4 and beyond. Let's not forget that there are still 3PD extensions which are still in the process of migrating to J4. Most sites can't migrate until it's clear which extensions can finish their migration and which are left behind to die.

Right now there are servers which fail to download the Joomla update package. If there is a security issue with Joomla 3 it will be very hard (if not impossible, depending on the site owner's skill level) to update these sites, exposing them to a security risks. We don't even provide an alternative method, like we do in Joomla 4 (you can update J4 over CLI where waiting even for half an hour for the update package to download is not a practical problem!). Therefore it makes sense to start by addressing this issue on Joomla 3 even though it's currently on security maintenance only.

If this approach works for y'all (and the production leadership ultimately decides that they cannot use CloudFlare as the default mirror which would solve this problem on most servers!) I can port this to Joomla 4.

### Testing Instructions

**Common instructions**

* Install a new site
* Apply the patch
* Change the `manifest_cache` in the `#__extensions` table for `extension_id` 700 to have a fake version, e.g. 3.10.10.
* Change the `libraries/src/Version.php` with PATCH_VERSION = 10
* Log into the site's backend
* Go to Components, Joomla Update

**Test 1**

Makes sure that updates without any further configuration still work the same way as they did before.

* Follow the Common instructions
* Click on Check for Updates
* An update is found
* Click on the button to install the update
* You wait for a while, the progress bar moves to 100% and the update installation starts.

**Test 2**

Makes sure that chunked downloads work.

* Follow the Common instructions
* Click on Options
* Click on Fine-tuning
* Set Chunked Downloads to Yes
* Click on Save & Close
* Click on Check for Updates
* An update is found
* Click on the button to install the update
* You will see the progress bar move every 3 seconds or more
* Once the download is complete the progress bar shows 100%, it becomes green and the update installation starts.

**Test 3**

Makes sure the fine-tuning setting for the minimum time works.

* Follow the Common instructions
* Click on Options
* Click on Fine-tuning
* Set Chunked Downloads to Yes
* Set “Minimum time between download steps (seconds)” to 10
* Click on Save & Close
* Click on Check for Updates
* An update is found
* Click on the button to install the update
* You will see the progress bar move every **10** seconds or more. The download is substantially slower than test 2.
* Once the download is complete the progress bar shows 100%, it becomes green and the update installation starts.

You can redo this test with the minimum time set to 0 and observe that the download is much faster — unless your server throws an error because of too many requests…

### Documentation Changes Required

Add the following to the help page for Joomla Update options.

The Fine-tuning tab's options control how Joomla Update will download the update packages before installing them on your site.

By default, Joomla Update tries to download the entire update package in a single page load. If your server has a slow network connection to the Joomla Project's update download servers it may time out before the download is complete, making it impossible to update your site.

If this happens to you — and _only_ if it happens! — you should change the “Chunked Downloads” option to Yes. Joomla Update will now download 1 megabyte of the update package at a time, minimising the risk of your server timing out.

The “Minimum time between download steps (seconds)” controls how fast these 1 megabyte downloads will come. If the last attempt to download 1 megabyte of the update takes less time than this setting is configured to, Joomla Update will wait for the remainder of the time. Setting this to 0 will make downloads very fast but there is a chance that your hosting company's server protection may block you and break the updates. If you receive an error from your server while the download is in progress increase this setting. If you do not see any errors try decreasing this setting to make downloads faster. The default value of 3 seconds strikes a good balance between download speed and working around server protections which would block the update package download on most commercial hosting companies.

### Translation Changes Required 

This PR introduces 10 new language strings.

### Important notes

This is for @HLeithner mainly and to explain why this option is disabled by default.

Every Joomla update provides a set of mirrors (alternate download URLs). The first mirror is the Joomla downloads site which runs a very old version of Akeeba Release System and is linked to an Amazon S3 bucket.

The items in this bucket are stored with Private ACL. ARS on the downloads site returns a [pre-signed URL](https://docs.aws.amazon.com/AmazonS3/latest/userguide/using-presigned-url.html) which allows only one specific HTTP verb: GET. There is no way to allow alternate verbs, such as HEAD.

To make chunked downloads possible we need to do a HEAD request on the mirror's download URL and follow all redirections. When we find the final destination we perform a final HEAD request and expect to see the HTTP Content-Length header which tells us how big is the file to download. Since the first mirror redirects to S3 using a pre-signed URL and the pre-signed URL does not allow HEAD requests we _can not use the default mirror for chunked downloads_. Therefore we will fall back to the next mirror which is GitHub and, if I recall correctly, GitHub wasn't very happy with Joomla using it for all its downloads.

In an attempt to alleviate the pain, I have implemented a simple workaround. If the user has NOT opted into chunked uploads I do NOT perform any HEAD requests. Instead, I simply get the list of mirror URLs and try them one after another, until one successfully downloads the entire file. Therefore the majority of users who are still on the single part download will be able to use the default mirror (hosted on S3) and all is good in the world.

If you want to make chunked downloads the default or only option you MUST do one of the following:
* Set the default mirror to the CloudFlare one. CloudFlare does not have bandwidth caps and does report the Content-Length correctly. Moreover, it has edge nodes in a huge number of locations, meaning that the download will be faster. It also does not hit the S3 bucket which means that it's also _cheaper_. Finally, if you use the Always Online feature in CloudFlare you can make sure that the update downloads work even when the Joomla server is down. Like, seriously, if you have CloudFlare why are you even using anything else?!
* If you do not want to use CloudFlare (why?!) you **MUST** set the ACLs of all download items on the S3 bucket to Public and migrate all `s3://` pseudo-URLs in the database to `http://` or `https://` URLs pointing directly to the file on the S3 bucket and change the item type to `link`. While certainly possible, I do not see why bother when you have a cheaper, faster, more reliable method.
